tools/generators/lib/PBXProj/src/Identifiers.swift

```diff
@@ -81,6 +83,161 @@ FF01000000000000000001\#(String(format: "%02X", index)) \#
+            hashCache: inout [UInt8: Set<String>]
+        ) -> SubIdentifier {
+            precondition(shard < 0xFE)
+            return SubIdentifier(
+                shard: String(format: "%02X", shard),
+                type: type,
+                path: path,
+                hash: shardSubIdentifier(
+                    path.path + (path.isFolder ? "0" : "1"),
+                    hashCache: &hashCache[shard, default: []]
+                )
+            )
+        }
+
+        public static func compileStubSubIdentifier(
+            targetSubIdentifier: Targets.SubIdentifier
+        ) -> SubIdentifier {
+            // We purposely store less information here than in
+            // `subIdentifier()`, to reduce the amount of data written into
+            // BuildFileMaps
+            return SubIdentifier(
+                shard: targetSubIdentifier.shard,
+                type: .compileStub,
+                path: BazelPath(""),
+                hash: targetSubIdentifier.hash
+            )
+        }
+
+        public static func id(
+            subIdentifier: SubIdentifier
+        ) -> String {
+            let type = subIdentifier.type
+
+            switch type {
+            case .compileStub:
+                return #"""
+\#(subIdentifier.shard)00\#(subIdentifier.hash)0000000000FF \#
+/* _CompileStub_.m in Sources */
+"""#
+
+            default:
+                let basename = subIdentifier.path.path
+                    .split(separator: "/").last!
+                return #"""
+\#(subIdentifier.shard)FF\#(subIdentifier.hash) \#
+/* \#(basename) in \#(type.buildPhase.name) */
+"""#
+            }
+        }
+
+        private static func shardSubIdentifier(
+            _ hashable: String,
+            hashCache: inout Set<String>
+        ) -> String {
+            var hash: String
+            var retryCount = 0
+            repeat {
+                hash = buildFileHash(hashable, retryCount: retryCount)
+                retryCount += 1
+            } while hashCache.contains(hash)
+
+            hashCache.insert(hash)
+
+            return hash
+        }
+
+        private static func buildFileHash(
+            _ hashable: String,
+            retryCount: Int
+        ) -> String {
+            let content: String
+            if retryCount == 0 {
+                content = hashable
+            } else {
+                content = "\(hashable)\0\(retryCount)"
+            }
+
+            let digest = Insecure.MD5.hash(data: content.data(using: .utf8)!)
+            return digest
+                // Xcode identifiers are 24 characters. We are using 4
+                // characters at the front. That leaves 20 characters that we
+                // can use. MD5 digests are 16 bytes (32 characters)
+                // long. So we need to truncate it to fit within the remaining
+                // 20 characters (by dropping 6 bytes). We choose the front 20
+                // because they are the most unique.
```

Comment:
> We choose the front 20 because they are the most unique.

Definitely a nit but aren't all parts of a cryptographic hash uniformly unique?

Reply:
I thought I read somewhere that it impacted the entropy, but now I'm reading the opposite. I'll remove that verbiage.